plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,208 @@
+
+data Log = LogDummy
+
+instance Pretty Log where
+    pretty = \case
+        LogDummy -> "TODO(@linj) remove this dummy log"
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor _recorder pluginId =
+    (defaultPluginDescriptor pluginId "Provides signature help of something callable")
+        { Ide.Types.pluginHandlers = mkPluginHandler SMethod_TextDocumentSignatureHelp signatureHelpProvider
+        }
+
+-- TODO(@linj) get doc
+signatureHelpProvider :: PluginMethodHandler IdeState Method_TextDocumentSignatureHelp
+signatureHelpProvider ideState _pluginId (SignatureHelpParams (TextDocumentIdentifier uri) position _mProgreeToken _mContext) = do
+    nfp <- getNormalizedFilePathE uri
+    mResult <- runIdeActionE "signatureHelp" (shakeExtras ideState) $ do
+        -- TODO(@linj) why HAR {hieAst} may have more than one AST?
+        (HAR {hieAst, hieKind}, positionMapping) <- useWithStaleFastE GetHieAst nfp
+        case fromCurrentPosition positionMapping position of
+            Nothing -> pure Nothing
+            Just oldPosition -> do
+                let functionName =
+                        extractInfoFromSmallestContainingFunctionApplicationAst
+                            oldPosition
+                            hieAst
+                            (\span -> getLeftMostNode >>> getNodeName span)
+                    functionType =
+                        extractInfoFromSmallestContainingFunctionApplicationAst
+                            oldPosition
+                            hieAst
+                            (\span -> getLeftMostNode >>> getNodeType hieKind span)
+                    argumentNumber =
+                        extractInfoFromSmallestContainingFunctionApplicationAst
+                            oldPosition
+                            hieAst
+                            getArgumentNumber
+                pure $ Just (functionName, functionType, argumentNumber)
+    case mResult of
+        -- TODO(@linj) what do non-singleton lists mean?
+        Just (functionName : _, functionType : _, argumentNumber : _) -> do
+            pure $ InL $ mkSignatureHelp functionName functionType (fromIntegral argumentNumber - 1)
+        _ -> pure $ InR Null
+
+mkSignatureHelp :: Name -> Text -> UInt -> SignatureHelp
+mkSignatureHelp functionName functionType argumentNumber =
+    let functionNameLabelPrefix = printOutputable (ppr functionName) <> " :: "
+     in SignatureHelp
+            [ SignatureInformation
+                  (functionNameLabelPrefix <> functionType)
+                  Nothing
+                  (Just $ mkArguments (fromIntegral $ T.length functionNameLabelPrefix) functionType)
+                  (Just $ InL argumentNumber)
+            ]
+            (Just 0)
+            (Just $ InL argumentNumber)
+
+-- TODO(@linj) can type string be a multi-line string?
+mkArguments :: UInt -> Text -> [ParameterInformation]
```

Comment:
I think this would be a lot simpler if you were just looking at the type itself!

Reply:
Although hmm, the offsets are into the string :scream: So maybe you would need to incrementally build up the `ParameterInformation`s and the overall `label`, but that would probably still be nicer?